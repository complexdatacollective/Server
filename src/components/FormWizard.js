import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button, Icon } from 'network-canvas-ui';

const prevButton = onClick => (
  <button key="prev" className="form__prev-button" aria-label="Submit" onClick={onClick}>
    <Icon name="form-arrow-left" />
  </button>
);

const nextButton = onClick => (
  <button key="next" className="form__next-button" aria-label="Submit" onClick={onClick}>
    <Icon name="form-arrow-right" />
  </button>
);

const submitButton = onClick => (
  <Button
    key="submit"
    size="small"
    content="Submit"
    onClick={onClick}
  />
);

class FormWizard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      pageIndex: 0,
    };
  }

  onSubmit = () => {
    this.props.onSubmit();
  }

  nextPage = () => {
    const count = this.props.children.length;
    this.setState({
      pageIndex: (this.state.pageIndex + 1 + count) % count,
    });
  };

  previousPage = () => {
    const count = this.props.children.length;
    this.setState({
      pageIndex: (this.state.pageIndex - 1 + count) % count,
    });
  };

  shouldShowNextButton = () => {
    const showingLastField = this.state.pageIndex === this.props.children.length - 1;
    return !showingLastField;
  }

  currentPage() {
    return this.props.children[this.state.pageIndex];
  }

  render() {
    return (
      <div className="form-wizard">
        {this.currentPage()}
        <div className="form-wizard__controls">
          <div className="form-wizard__previous">
            {this.state.pageIndex !== 0 && this.props.prevButton(this.previousPage)}
          </div>
          <div className="form-wizard__progress">
            Step {this.state.pageIndex + 1} of {this.props.children.length}
          </div>
          <div className="form-wizard__next">
            {this.shouldShowNextButton() ?
              this.props.nextButton(this.nextPage) :
              this.props.submitButton(this.onSubmit)
            }
          </div>
        </div>
      </div>
    );
  }
}

FormWizard.propTypes = {
  children: PropTypes.node.isRequired,
  prevButton: PropTypes.func,
  nextButton: PropTypes.func,
  submitButton: PropTypes.func,
  onSubmit: PropTypes.func,
};

FormWizard.defaultProps = {
  prevButton,
  nextButton,
  submitButton,
  onSubmit: () => {},
};

export default FormWizard;
