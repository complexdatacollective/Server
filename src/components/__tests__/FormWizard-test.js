/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import FormWizard from '../FormWizard';

const nextButton = onClick => (<button onClick={onClick}>next</button>);
const prevButton = onClick => (<button onClick={onClick}>prev</button>);
const mockProps = {
  nextButton,
  prevButton,
};

describe('<FormWizard />', () => {
  it('should render', () => {
    const subject = shallow((
      <FormWizard {...mockProps}>
        <div>Slide 1</div>
        <div>Slide 2</div>
      </FormWizard>
    ));

    expect(subject).toMatchSnapshot();
  });
});
