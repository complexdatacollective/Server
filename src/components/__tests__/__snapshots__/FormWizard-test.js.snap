// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormWizard /> should render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormWizard
    nextButton={[Function]}
    onSubmit={[Function]}
    prevButton={[Function]}
    submitButton={[Function]}
>
    <div>
        Slide 1
    </div>
    <div>
        Slide 2
    </div>
</FormWizard>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          Slide 1
</div>,
        <div
          className="form-wizard__controls"
>
          <div
                    className="form-wizard__previous"
          />
          <div
                    className="form-wizard__progress"
          >
                    Step 
                    1
                     of 
                    2
          </div>
          <div
                    className="form-wizard__next"
          >
                    <button
                              onClick={[Function]}
                    >
                              next
                    </button>
          </div>
</div>,
      ],
      "className": "form-wizard",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Slide 1",
        },
        "ref": null,
        "rendered": "Slide 1",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="form-wizard__previous"
/>,
            <div
              className="form-wizard__progress"
>
              Step 
              1
               of 
              2
</div>,
            <div
              className="form-wizard__next"
>
              <button
                            onClick={[Function]}
              >
                            next
              </button>
</div>,
          ],
          "className": "form-wizard__controls",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": false,
              "className": "form-wizard__previous",
            },
            "ref": null,
            "rendered": false,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Step ",
                1,
                " of ",
                2,
              ],
              "className": "form-wizard__progress",
            },
            "ref": null,
            "rendered": Array [
              "Step ",
              1,
              " of ",
              2,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <button
                onClick={[Function]}
>
                next
</button>,
              "className": "form-wizard__next",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "next",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "next",
              "type": "button",
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            Slide 1
</div>,
          <div
            className="form-wizard__controls"
>
            <div
                        className="form-wizard__previous"
            />
            <div
                        className="form-wizard__progress"
            >
                        Step 
                        1
                         of 
                        2
            </div>
            <div
                        className="form-wizard__next"
            >
                        <button
                                    onClick={[Function]}
                        >
                                    next
                        </button>
            </div>
</div>,
        ],
        "className": "form-wizard",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Slide 1",
          },
          "ref": null,
          "rendered": "Slide 1",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="form-wizard__previous"
/>,
              <div
                className="form-wizard__progress"
>
                Step 
                1
                 of 
                2
</div>,
              <div
                className="form-wizard__next"
>
                <button
                                onClick={[Function]}
                >
                                next
                </button>
</div>,
            ],
            "className": "form-wizard__controls",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": false,
                "className": "form-wizard__previous",
              },
              "ref": null,
              "rendered": false,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Step ",
                  1,
                  " of ",
                  2,
                ],
                "className": "form-wizard__progress",
              },
              "ref": null,
              "rendered": Array [
                "Step ",
                1,
                " of ",
                2,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <button
                  onClick={[Function]}
>
                  next
</button>,
                "className": "form-wizard__next",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "next",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "next",
                "type": "button",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
